### ISSUE TITLE
darktable-cltest fails on latest Fedora 37 with AMD RX 6800 and OpenCL enabled

### ISSUE THREAD

**Issue by starsforeveryone** [darktable-org/darktable#12971]
Hello dear developers,

I've just noticed that I cannot enable OpenCL in Darktable on my new computer that I built, since the checkboxes necessary to do so are all greyed out.

I assume this is due to darktable-cltest failing. When running the command on the terminal, this is the output:

```
[dt_get_sysresource_level] switched to 2 as `large'
  total mem:       64036MB
  mipmap cache:    8004MB
  available mem:   43774MB
  singlebuff:      1000MB
  OpenCL tune mem: OFF
  OpenCL pinned:   OFF
[opencl_init] opencl related configuration options:
[opencl_init] opencl: ON
[opencl_init] opencl_scheduling_profile: 'default'
[opencl_init] opencl_library: 'default path'
[opencl_init] opencl_device_priority: '*/!0,*/*/*/!0,*'
[opencl_init] opencl_mandatory_timeout: 400
[opencl_init] opencl_synch_cache: active module
[opencl_init] opencl library 'libOpenCL.so.1' found on your system and loaded
[opencl_init] found 1 platform
[opencl_init] found 1 device

[dt_opencl_device_init]
   DEVICE:                   0: 'gfx1030'
   CANONICAL NAME:           gfx1030
   PLATFORM NAME & VENDOR:   AMD Accelerated Parallel Processing, Advanced Micro Devices, Inc.
   DRIVER VERSION:           3486.0 (HSA1.1,LC)
   DEVICE VERSION:           OpenCL 2.0 
   DEVICE_TYPE:              GPU
   GLOBAL MEM SIZE:          16368 MB
   MAX MEM ALLOC:            13913 MB
   MAX IMAGE SIZE:           16384 x 16384
   MAX WORK GROUP SIZE:      256
   MAX WORK ITEM DIMENSIONS: 3
   MAX WORK ITEM SIZES:      [ 1024 1024 1024 ]
   ASYNC PIXELPIPE:          NO
   PINNED MEMORY TRANSFER:   NO
   MEMORY TUNING:            NO
   FORCED HEADROOM:          400
   AVOID ATOMICS:            NO
   MICRO NAP:                250
   ROUNDUP WIDTH:            16
   ROUNDUP HEIGHT:           16
   CHECK EVENT HANDLES:      128
   DEFAULT DEVICE:           NO
   KERNEL DIRECTORY:         /usr/share/darktable/kernels
   CL COMPILER OPTION:       -cl-fast-relaxed-math
0.205922 [opencl_fopen_stat] could not open file `/home/xxxxxxxx/.cache/darktable/cached_kernels_for_gfx1030_34860HSA11LC/demosaic_ppg.cl.bin'!
0.205930 [opencl_load_program] could not load cached binary program, trying to compile source
0.215375 [opencl_build_program] could not build program: CL_BUILD_PROGRAM_FAILURE
0.215382 [dt_opencl_device_init] failed to compile program `demosaic_ppg.cl'!
[opencl_init] no suitable devices found.
[opencl_init] FINALLY: opencl is NOT AVAILABLE on this system.
[opencl_init] initial status of opencl enabled flag is OFF.
```

I'm running an AMD Ryzen 7950X with 64 GiB DDR5-RAM and an AMD RX 6800 16GB on the freshly released Fedora 37.

Could this be a bug in Darktable?

Thanks in advance!

**Comment by garrett** [comment]
I'm hitting this on an AMD RX 6700XT as well. It worked on stable darktable in Fedora 37 for me before, but it's not working any more on nightly darktable. (Perhaps I did have things cached before and it's just failing with compilation?)

**Comment by garrett** [comment]
I wonder if it's related to the new build of rocm-opencl from last week @ https://bodhi.fedoraproject.org/updates/FEDORA-2022-e16456fe6b which landed in stable yesterday?

**Comment by starsforeveryone** [comment]
Thank you very much for investigating! :smile:
If I can be of any help, let me know.

**Comment by starsforeveryone** [comment]
Just tried to delete ~/.cache/darktable and re-run darktable-cltest, sadly, still the same error message. :(

I'd really love to use OpenCL in darktable!

**Comment by starsforeveryone** [comment]
Any update on this?

**Comment by oleastre** [comment]
I had the same error on my Fedora 37 system, and looked at the darktable code.
If you enable verbose debug messages, you should find the reason of the `CL_BUILD_PROGRAM_FAILURE`.
what I did:
```
$ darktable-cltest -d verbose
0.373041 [opencl_load_program] successfully loaded program from '/opt/darktable-dev/share/darktable/kernels/demosaic_ppg.cl' MD5: 'c9a671586912a40b7872cfc9b82fd905'
0.385586 [opencl_build_program] could not build program: CL_BUILD_PROGRAM_FAILURE
0.385600 [opencl_build_program] BUILD STATUS: -2
0.385603 BUILD LOG:
0.385604 fatal error: malformed or corrupted AST file: 'could not find file '/usr/lib64/clang/15.0.0/include/opencl-c-base.h' referenced by AST file '/tmp/comgr-669302/include/opencl1.2-c.pch''
1 error generated.
Error: Failed to compile source (from CL or HIP source to LLVM IR).
```
As you can see there, there is a missing `opencl-c-base.h` file.
I searched for it in the fedora archive: 
```
$ dnf provides /usr/lib64/clang/15.0.0/include/opencl-c-base.h
Last metadata expiration check: 0:36:08 ago on Thu Dec 15 00:01:09 2022.
clang-libs-15.0.0-2.fc37.x86_64 : Runtime library for clang
Repo        : @System
Matched from:
Filename    : /usr/lib64/clang/15.0.0/include/opencl-c-base.h

clang-libs-15.0.0-2.fc37.x86_64 : Runtime library for clang
Repo        : fedora
Matched from:
Filename    : /usr/lib64/clang/15.0.0/include/opencl-c-base.h
```
I just installed that `clang-libs` package, and restarted `darktable-cltest` and it successfully compiled all kernels.

I suspect this is a fedora packaging issue, they probably miss the dependency on llvm that is required to compile the opencl code

**Comment by starsforeveryone** [comment]
Thank you very much dude, this saved me so much trouble! oO Awesome!

I could easily reproduce what you discovered:
`darktable-cltest -d verbose`
yielded the same error you got:
`0.215396 fatal error: malformed or corrupted AST file: 'could not find file '/usr/lib64/clang/15.0.0/include/opencl-c-base.h' referenced by AST file '/tmp/comgr-854f4c/include/opencl1.2-c.pch''
`

Now here comes the funny part: It get's even weirder!
`sudo dnf install clang-libs`
told me the package was already installed?

Quick reinstall:
`sudo dnf reinstall clang-libs`
`Reinstalled:
  clang-libs-15.0.6-1.fc37.x86_64                                                                                                                                                                                                                                                                                                                                                                                                                                                             
`
Same error!

Wait a minute...
`sudo dnf provides /usr/lib64/clang/15.0.0/include/opencl-c-base.h`
gave me back
`clang-libs-15.0.0-2.fc37.x86_64 : Runtime library for clang`

🧐

Why does it say _clang-libs-15.0.0_-2.fc37 when I had _clang-libs-15.0.6_-1.fc37 installed? Oh could it be...

`sudo dnf install clang-libs-15.0.0-2.fc37.x86_64`
==> Yes, this one was NOT installed!

Then I ran darktable-cltest again:
`darktable-cltest`
resulted in:
`[opencl_init] here are the internal numbers and names of OpenCL devices available to darktable:
[opencl_init]		0	'gfx1030'
**[opencl_init] FINALLY: opencl is AVAILABLE on this system.**
`

I know this line was not intended as a joke by the programmers 🤣 but it really made me laugh. FINALLY, OpenCL is available on my system! 😎

TLDR:
Fedora is already shipped with a newer version of _clang-libs_, that's probably why dnf didn't catch the dependency problem.

Lastly, if somebody could report this as a bug to Fedora's darktable maintainer, it would be really appreciated to help those who aren't as lucky to find this here.

Thanks very much to @oleastre & all the other people who participated!

**Comment by garrett** [comment]
> Lastly, if somebody could report this as a bug to Fedora's darktable maintainer

Isn't this a problem with clang-libs in Fedora and not darktable itself? It's the package that seems to mistakenly not ship the necessary file in the latest minor version in Fedora.

(I can't get OpenCL working with darktable from OBS either, for example.)

Filed against clang in Fedora @ https://bugzilla.redhat.com/show_bug.cgi?id=2154666

**Comment by garrett** [comment]
Also mentioned @ https://bodhi.fedoraproject.org/updates/FEDORA-2022-a61fda3173, hoping to raise awareness of this issue.
